File: backend/src/main/java/com/library/service/LoanService.java
package com.library.service;

import com.library.dto.LoanDTO;

import java.util.List;

/**
 * Інтерфейс для роботи з позиками книг (Interface Segregation Principle)
 */
public interface LoanService {

    LoanDTO createLoan(Long userId, Long bookId);

    LoanDTO getLoanById(Long id);

    List<LoanDTO> getLoansByUserId(Long userId);

    List<LoanDTO> getLoansByBookId(Long bookId);

    List<LoanDTO> getAllLoans();

    List<LoanDTO> getOverdueLoans();

    LoanDTO returnBook(Long loanId);

    LoanDTO renewLoan(Long loanId);

    void updateOverdueLoans();
}

File: frontend/src/services/loanService.js
import api from './api';

const loanService = {
  getAllLoans: () => api.get('/loans'),

  getLoanById: (id) => api.get(`/loans/${id}`),

  getLoansByUser: (userId) => api.get(`/loans/user/${userId}`),

  getLoansByBook: (bookId) => api.get(`/loans/book/${bookId}`),

  getOverdueLoans: () => api.get('/loans/overdue'),

  createLoan: (userId, bookId) => api.post('/loans', null, { params: { userId, bookId } }),

  returnBook: (loanId) => api.put(`/loans/${loanId}/return`),

  renewLoan: (loanId) => api.put(`/loans/${loanId}/renew`),
};

export default loanService;
